src/relay/backend/vm/manifest_lifetimes.cc

```diff
@@ -0,0 +1,611 @@
+        succ.push_back(parent->nodes[index + 1]);
+      }
+      return succ;
+    }
+
+    /*! \brief Creates a node with the given expr and appends it to the parent basic block. */
+    static NodePtr Make(BasicBlockPtr parent, Expr expr) {
+      NodePtr n = std::make_shared<Node>();
+      n->parent = parent;
+      n->expr = expr;
+      n->index = parent->nodes.size();
+      parent->nodes.push_back(n);
+      return n;
+    }
+  };
+
+  /*! \brief The basic block where control flow begins. */
+  BasicBlockPtr entry;
+
+  /*!
+   * \brief Mapping from Let expressions to their corresponding nodes. Note that Let expressions
+   * are never shared in ANF (unlike vars), so this is an injection.
+   */
+  std::unordered_map<Expr, NodePtr, ObjectPtrHash, ObjectPtrEqual> let_map;
+
+  /*! \brief The nodes of the CFG in reverse post order. */
+  std::vector<NodePtr> reverse_post_order;
+
+  /*! \brief Creates and returns the CFG of the given expression. */
+  static ControlFlowGraph Create(const Expr& body);
+
+ private:
+  class Creator;
+};
+
+/*! \brief Helper class for building CFGs. */
+class ControlFlowGraph::Creator : private ExprFunctor<void(const Expr&, BasicBlockPtr)> {
+ public:
+  Creator() {}
+
+  ControlFlowGraph Create(const Expr& body) {
+    cfg_.entry = BasicBlock::Make();
+    VisitExpr(body, cfg_.entry);
+    return std::move(cfg_);
+  }
+
+ private:
+  /*! \brief The CFG being built. */
+  ControlFlowGraph cfg_;
+  /*!
+   * \brief Whether or not we are in a function. CFGs do not support nested functions so this is
+   * used to error out in such a case.
+   */
+  bool in_func_ = false;
+
+  /*!
+   * \brief Link \p to as a successor block to \p from.
+   */
+  void Succ(BasicBlockPtr from, BasicBlockPtr to) {
+    from->succ.push_back(to);
+    to->pred.push_back(from);
+  }
+
+#define DEFAULT_CFG(OP)                                       \
+  void VisitExpr_(const OP* op, BasicBlockPtr parent) final { \
+    NodePtr n = Node::Make(parent, GetRef<Expr>(op));         \
+    cfg_.reverse_post_order.push_back(n);                     \
+  }
+
+  void VisitExpr_(const FunctionNode* f, BasicBlockPtr parent) final {
+    ICHECK(!in_func_) << "nested functions not supported by CFG analysis";
+    in_func_ = true;
+
+    // Unwrap the nested function and proceed normally.
+    if (f->HasNonzeroAttr(attr::kClosure)) {
+      ICHECK(f->body.as<FunctionNode>());
+      return VisitExpr(Downcast<Function>(f->body)->body, parent);
+    }
+
+    return VisitExpr(f->body, parent);
+  }
+
+  void VisitExpr_(const LetNode* let_node, BasicBlockPtr parent) final {
+    Expr expr = GetRef<Expr>(let_node);
+
+    while (const LetNode* inner_let_node = expr.as<LetNode>()) {
+      NodePtr curr_node = Node::Make(parent, expr);
+
+      ICHECK(!cfg_.let_map.count(expr));
+      cfg_.let_map[expr] = curr_node;
+      cfg_.reverse_post_order.push_back(curr_node);
+
+      if (const IfNode* ite = AsIgnoringOnDevice<IfNode>(inner_let_node->value)) {
```

Comment:
nit: Reminder comment that we are dealing with the 'exits' from the bb, hence only looking at the two branch exprs

Reply:
let me know if the comment I added helps at all